components/eamxx/src/control/intensive_observation_period.cpp

```diff
@@ -237,6 +495,354 @@ read_fields_from_file_for_iop (const std::string& file_name,
+  bool has_level_data = false;
+  for (auto& it : m_iop_fields) {
+    if (it.second.rank() == 1) {
+      has_level_data = true;
+      break;
+    }
+  }
+
+  // Compute values and indices associate with pressure for interpolating data (if necessary).
+  int adjusted_file_levs;
+  int iop_file_start;
+  int iop_file_end;
+  int model_start;
+  int model_end;
+  if (has_level_data) {
+    // Load surface pressure (Ps) from iop file
+    auto ps_data = surface_pressure.get_view<Real, Host>().data();
+    read_variable_from_file(iop_file, "Ps", "real", {"lon","lat"}, iop_file_time_idx, ps_data);
+    surface_pressure.sync_to_dev();
+
+    // Pre-process file pressures, store number of file levels
+    // where the last level is the first level equal to surface pressure.
+    const auto iop_file_pres_v = iop_file_pressure.get_view<Real*>();
+    // Sanity check
+    EKAT_REQUIRE_MSG(file_levs+1 == iop_file_pres_v.extent_int(0),
+                    "Error! Unexpected size for helper field \"iop_file_pressure\"\n");
+    const auto& Ps = surface_pressure.get_view<const Real>();
+    Kokkos::parallel_reduce(file_levs+1, KOKKOS_LAMBDA (const int ilev, int& lmin) {
+      if (ilev == file_levs) {
+        // Add surface pressure to last iop file pressure entry
+        iop_file_pres_v(ilev) = Ps()/100;
+      }
+      if (iop_file_pres_v(ilev) > Ps()/100) {
+        // Set upper bound on pressure values
+        iop_file_pres_v(ilev) = Ps()/100;
+      }
+      if (iop_file_pres_v(ilev) == Ps()/100) {
+        // Find minimum number of levels where the final
+        // level would contain the largest value.
+        if (ilev < lmin) lmin = ilev+1;
+      }
+    }, Kokkos::Min<int>(adjusted_file_levs));
+
+    EKAT_REQUIRE_MSG(adjusted_file_levs > 1,
+                     "Error! Pressures in iop file "+iop_file+" is are inccorrectly set. "
+                     "Surface pressure \"Ps\" (converted to milibar) should be greater "
+                     "than at least the 1st entry in midpoint pressures \"lev\".\n");
+
+    // Compute model pressure levels
+    const auto model_pres_v = model_pressure.get_view<Real*>();
+    const auto model_nlevs = model_pres_v.extent(0);
+    const auto hyam_v = m_helper_fields["hyam"].get_view<const Real*>();
+    const auto hybm_v = m_helper_fields["hybm"].get_view<const Real*>();
+    Kokkos::parallel_for(model_nlevs, KOKKOS_LAMBDA (const int ilev) {
+      model_pres_v(ilev) = 1000*hyam_v(ilev) + Ps()*hybm_v(ilev)/100;
+    });
+
+    // Find file pressure levels just outside the range of model pressure levels
+    Kokkos::parallel_reduce(adjusted_file_levs, KOKKOS_LAMBDA (const int& ilev, int& lmax, int& lmin) {
+      if (iop_file_pres_v(ilev) <= model_pres_v(0) && ilev > lmax) {
+        lmax = ilev;
+      }
+      if (iop_file_pres_v(ilev) >= model_pres_v(model_nlevs-1) && ilev+1 < lmin) {
+        lmin = ilev+1;
+      }
+    },
+    Kokkos::Max<int>(iop_file_start),
```

Comment:
I was not aware that Kokkos allowed to pass 2+ elements to be reduced. Nice.

Reply:
Yeah, I was unsure if that was in the version of Kokkos that we are using, but turns out it is!